metrics/graphite/graphite.go

```diff
@@ -0,0 +1,295 @@
+// should be integers in the range 1..99. The gauge names are assigned by
+// using the passed name as a prefix and appending "_pNN" e.g. "_p50".
+//
+// The values of this histogram will be periodically emitted in a Graphite-compatible
+// format once the Emitter is started. Fields are ignored.
+func (e *emitter) NewHistogram(name string, minValue, maxValue int64, sigfigs int, quantiles ...int) metrics.Histogram {
+	// only one flush at a time
+	e.metricMu.Lock()
+	defer e.metricMu.Unlock()
+
+	gauges := map[int]metrics.Gauge{}
+	for _, quantile := range quantiles {
+		if quantile <= 0 || quantile >= 100 {
+			panic(fmt.Sprintf("invalid quantile %d", quantile))
+		}
+		gauges[quantile] = e.gauge(fmt.Sprintf("%s_p%02d", name, quantile))
+	}
+	h := newWindowedHistogram(name, minValue, maxValue, sigfigs, gauges)
+	e.histograms = append(e.histograms, h)
+	return h
+}
+
+// NewTimeHistogram returns a TimeHistogram wrapper around the windowed
+// HDR histrogram provided by this package.
+func (e *emitter) NewTimeHistogram(name string, unit time.Duration, minValue, maxValue int64, sigfigs int, quantiles ...int) metrics.TimeHistogram {
+	h := e.NewHistogram(name, minValue, maxValue, sigfigs, quantiles...)
+	return metrics.NewTimeHistogram(unit, h)
+}
+
+// NewGauge returns a Gauge whose value will be periodically emitted in
+// a Graphite-compatible format once the Emitter is started. Fields are ignored.
+func (e *emitter) NewGauge(name string) metrics.Gauge {
+	// only one flush at a time
+	e.metricMu.Lock()
+	defer e.metricMu.Unlock()
+	return e.gauge(name)
+}
+
+func (e *emitter) gauge(name string) metrics.Gauge {
+	g := &gauge{name, 0}
+	e.gauges = append(e.gauges, g)
+	return g
+}
+
+// Start will kick off a background goroutine to
+// call Flush once every interval.
+func (e *emitter) Start(interval time.Duration) {
+	go func() {
+		t := time.Tick(interval)
+		for range t {
+			err := e.Flush()
+			if err != nil {
+				log.Print("error: could not dial graphite host: ", err)
+				continue
+			}
+		}
+	}()
+}
+
+// Flush will attempt to create a connection with the given address
+// and write the current metrics to it in a Graphite-compatible format.
+//
+// Users can call this method on process shutdown to ensure
+// the current metrics are pushed to Graphite.
+func (e *emitter) Flush() error {
+	// open connection
+	conn, err := net.DialTCP("tcp", nil, e.addr)
+	if err != nil {
+		return err
+	}
+
+	// flush stats to connection
+	e.flush(conn)
+
+	// close connection
+	conn.Close()
+	return nil
+}
+
+func (e *emitter) flush(conn io.Writer) {
+	// only one flush at a time
+	e.metricMu.Lock()
+	defer e.metricMu.Unlock()
+
+	// buffer the writer and make sure to flush it
+	w := bufio.NewWriter(conn)
+	defer w.Flush()
+
+	now := time.Now().Unix()
+
+	// emit counter stats
+	for _, c := range e.counters {
+		fmt.Fprintf(w, "%s.%s.count %d %d\n", e.prefix, c.Name(), c.count, now)
```

Comment:
Can you drop a link to this format on e.g. the Graphite docs into a comment somewhere?